### PR TITLE
Do not allow creating registers with bitsize 0

### PR DESCRIPTION
**Pull request by tanujkhattar**
Registers with bitsize 0 are ambiguous since they you are not really passing any qubits to instantiate those registers. They also complicate the interop with Qualtran. This PR removes this workflow by adding an explicit validation that `bitsize` should always be a positive integer. 

cc @mpharrigan

### CODE REVIEW

File: cirq-ft/cirq_ft/algos/apply_gate_to_lth_target.py [hunk 1]
```diff
@@ -77,7 +77,7 @@ def selection_registers(self) -> Tuple[infra.SelectionRegister, ...]:
 
     @cached_property
     def target_registers(self) -> Tuple[infra.Register, ...]:
-        total_iteration_size = np.product(
+        total_iteration_size = np.prod(
```
Comment (mpharrigan) [inline cirq-ft/cirq_ft/algos/apply_gate_to_lth_target.py:80]: for my edification: why this change?

**Review by mpharrigan — APPROVED** [review]

File: cirq-ft/cirq_ft/algos/apply_gate_to_lth_target.py (hunk 1 above)
Comment (tanujkhattar) [reply cirq-ft/cirq_ft/algos/apply_gate_to_lth_target.py:80]: It fails mypy checks because `np.product` has been deprecated.
